src/api_server/src/attest/amber.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2023 by Intel.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::Attest;
+use anyhow::*;
+use as_types::AttestationResults;
+use async_trait::async_trait;
+use jsonwebtoken::{decode, decode_header, jwk, DecodingKey, Validation};
+use kbs_types::{Attestation, Tee};
+use reqwest::header::{ACCEPT, CONTENT_TYPE};
+use serde::{Deserialize, Serialize};
+use std::fs::File;
+use std::io::BufReader;
+
+#[derive(Deserialize, Debug)]
+struct TdxEvidence {
+    _cc_eventlog: Option<String>,
+    quote: String,
+}
+
+#[derive(Serialize, Debug)]
+struct AttestReqData {
+    quote: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct AttestRespData {
+    token: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct Claims {
+    amber_unmatched_policy_ids: Option<Vec<serde_json::Value>>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct AmberConfig {
+    pub base_url: String,
+    pub api_key: String,
+    pub certs_file: String,
+    pub check_policy: Option<bool>,
+}
+
+pub struct Amber {
+    config: AmberConfig,
+    jwks: jwk::JwkSet,
+}
+
+#[async_trait]
+impl Attest for Amber {
+    async fn attest_verify(
+        &mut self,
+        tee: Tee,
+        _nonce: &str,
+        attestation: &str,
+    ) -> Result<AttestationResults> {
+        if tee != Tee::Tdx {
+            return Err(anyhow!("Only implement for tdx now"));
+        }
+        // get quote
+        let attestation = serde_json::from_str::<Attestation>(attestation)
+            .map_err(|e| anyhow!("Deserialize Attestation failed: {:?}", e))?;
+        let tdx_evidence = serde_json::from_str::<TdxEvidence>(&attestation.tee_evidence)
+            .map_err(|e| anyhow!("Deserialize TDX Evidence failed: {:?}", e))?;
+
+        // construct attest request data
+        let req_data = AttestReqData {
+            quote: tdx_evidence.quote,
+        };
+
+        let attest_req_body = serde_json::to_string(&req_data)
+            .map_err(|e| anyhow!("Serialize attestation request body failed: {:?}", e))?;
+
+        // send attest request
+        log::info!("post attestation request ...");
+        let client = reqwest::Client::new();
+        let resp = client
+            .post(format!("{}/appraisal/v1/attest", &self.config.base_url))
+            .header(CONTENT_TYPE, "application/json")
+            .header(ACCEPT, "application/json")
+            .header("x-api-key", &self.config.api_key)
+            .body(attest_req_body)
+            .send()
+            .await
+            .map_err(|e| anyhow!("Post attestation request failed: {:?}", e))?;
+
+        if resp.status() != reqwest::StatusCode::OK {
+            return Err(anyhow!(
```

Comment:
We can use `bail!` macro to make it dry
```suggestion
            bail!("Attestation request failed: respone status={}", resp.status());
```

Reply:
I updated the patch.